design/14_tinkerbell_crd_refactor/proposal.md

```diff
@@ -0,0 +1,462 @@
+	// It is necessary to specify at least one NetworkInterface.
+	NetworkInterfaces NetworkInterfaces
+
+	// IPXE provides iPXE script override fields.
+	// Optional.
+	IPXE IPXE
+
+	// OSIE describes the Operating System Installation Environment to be netbooted.
+	OSIE OSIE
+
+	// Instance describes instance specific data that is generally unused by Tinkerbell core.
+	Instance Instance
+
+	// StorageDevices is a list of storage devices that will be available in the OSIE.
+	// Optional.
+	StorageDevices []StorageDevice
+
+	// BMCRef references a Rufio Machine object. It exists in the current API and will not be changed
+	// with this proposal.
+	BMCRef LocalObjectReference
+}
+
+// NetworkInterface is the desired configuration for a particular network interface.
+type NetworkInterface struct {
+	// DHCP is the basic network information for serving DHCP requests.
+	DHCP DHCP
+
+	// DisableDHCP disables DHCP for this interface. Implies DisableNetboot.
+	// Default false.
+	DisableDHCP bool
+
+	// DisableNetboot disables netbooting for this interface. The interface will still receive
+	// network information speified on by DHCP.
+	// Default false.
+	DisableNetboot bool
+}
+
+// DHCP describes basic network configuration to be served in DHCP offers.
+type DHCP struct {
+	IP      string
+	Netmask string
+
+	// Optional.
+	Gateway string
+
+	// Optional.
+	Hostname string
+
+	// Optional.
+	VLANID int
+
+	// Optional.
+	Nameservers []string
+
+	// Optional.
+	Timeservers []string
+
+	// Defaults to max allowed DHCP lease time.
+	LeaseTime int32
+}
+
+// OSIE describes an OSIE to be used with a Hardware. The environment data
+// is dependent on the OSIE being used and should be updated with the OSIE reference object.
+type Netboot struct {
+	// OSIERef is a reference to an OSIE object.
+	OSIERef LocalObjectReference
+
+	// KernelParams passed to the kernel when launching the OSIE. Parameters are joined with a
+	// space.
+	// Optional.
+	KernelParams []KernelParam
+}
+
+type IPXE struct {
+	// Inline is an inline iPXE script that will be served as specified on this property.
+	Inline *string
+
+	// URL is a URL to an hosted iPXE script.
+	URL *string
+}
+
+// Instance describes instance specific data. Instance specific data is typically dependent on the
+// permanent OS that a piece of hardware runs. This data is often served by an instance metadata
+// service such as Tinkerbell's Hegel. The core Tinkerbell stack does not leverage this data.
+type Instance struct {
+	// Userdata is data with a structure understood by the producer and consumer of the data.
+	Userdata string
+
+	// Vendordata is data with a structure understood by the producer and consumer of the data.
+	Vendordata string
+}
+
+// NetworkInterfaces maps a MAC address to a NetworkInterface.
+type NetworkInterfaces map[MAC]NetworkInterface
+
+// MAC is a Media Access Control address.
+type MAC string
```

Comment:
Why this field and the other fields below(`StorageDevice`,  `KernelParam`) should be a distinct types?

Reply:
1. So we can use CEL type validation.
2. Some are used as map keys and being explicit about what we expect is useful vs `string`.